File: ui-testing/pages/cart_page.py
from pages.base_page import BasePage
from selenium.webdriver.common.by import By


class Cart(BasePage):
    CONTINUE_SHOP = (By.ID, 'continue-shopping')
    CHECKOUT = (By.ID, 'checkout')
    CART_TITLE = (By.XPATH, "//span[@data-test='title']")

    CART_ITEM_CONTAINER = (By.CLASS_NAME, "cart_item")
    ITEM_NAME = (By.CLASS_NAME, "inventory_item_name")
    ITEM_PRICE = (By.CLASS_NAME, "inventory_item_price")
    ITEM_QTY = (By.CLASS_NAME, "cart_quantity")

    def get_all_items_detail(self):

        items_elements = self.find_elements(self.CART_ITEM_CONTAINER)
        parsed_items = []
        for element in items_elements:
            name = element.find_element(*self.ITEM_NAME).text
            price_raw = element.find_element(*self.ITEM_PRICE).text
            qty = element.find_element(*self.ITEM_QTY).text
            parsed_items.append({
                "name": name,
                "price": float(price_raw.replace('$', '')),
                "qty": int(qty)
            })
        return parsed_items

File: playwright-testing/pages/cart_page.ts
import { Page, Locator } from '@playwright/test';
import { BasePage } from './base.page';

export class CartPage extends BasePage {
    readonly continueShoppingButton: Locator;
    readonly checkoutButton: Locator;
    readonly cartTitle: Locator;
    readonly cartItemContainer: Locator;

    constructor(page: Page) {
        super(page);
        this.continueShoppingButton = page.locator('#continue-shopping');
        this.checkoutButton = page.locator('#checkout');
        this.cartTitle = page.locator("[data-test='title']");
        this.cartItemContainer = page.locator('.cart_item');
    }

    async getAllItemsDetail(): Promise<{ name: string; price: number; qty: number }[]> {
        const count = await this.cartItemContainer.count();
        const itemDetails = [];

        for (let i = 0; i < count; i++) {
            const item = this.cartItemContainer.nth(i);
            const name = await item.locator('.inventory_item_name').innerText();
            const priceRaw = await item.locator('.inventory_item_price').innerText();
            const qty = await item.locator('.cart_quantity').innerText();

            itemDetails.push({
                name,
                price: parseFloat(priceRaw.replace('$', '')),
                qty: parseInt(qty, 10),
            });
        }

        return itemDetails;
    }

    async clickContinueShopping(): Promise<void> {
        await this.continueShoppingButton.click();
    }

    async clickCheckout(): Promise<void> {
        await this.checkoutButton.click();
    }
}